File: android/app/src/test/kotlin/run/nuri/getagrip/runner/RunnerActivityLifecycleTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.runner

import run.nuri.getagrip.FakeClock
import run.nuri.getagrip.RecordingProgressorClient
import run.nuri.getagrip.ble.ProgressorConnectionState
import run.nuri.getagrip.ble.StreamStartCause
import run.nuri.getagrip.engine.ForceSample
import run.nuri.getagrip.engine.HandMode
import run.nuri.getagrip.engine.RunnerEvent
import run.nuri.getagrip.engine.RunnerPhase
import run.nuri.getagrip.engine.SessionPlan
import run.nuri.getagrip.engine.SetPlan
import run.nuri.getagrip.inertScope
import run.nuri.getagrip.store.DeviceStore
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class RunnerActivityLifecycleTests {
    private class ActivityRecorder : ActivityPublisher {
        override var isRunning = false
        var starts = 0
        var ends = 0
        val states = mutableListOf<SessionActivityState>()
        override fun start(routineName: String, plannedReps: Int, setCount: Int, state: SessionActivityState) {
            starts++
            isRunning = true
            states += state
        }
        override fun update(state: SessionActivityState) { states += state }
        override fun end() {
            if (!isRunning) return
            isRunning = false
            ends++
        }
    }
    private class ServiceRecorder : SessionServiceController {
        var starts = 0
        var ends = 0
        override fun begin() { starts++ }
        override fun end() { ends++ }
    }

    @Test fun releaseRestPauseAndFinishPublishTheirActualLifecycle() {
        val client = RecordingProgressorClient()
        val device = DeviceStore(client, scope = inertScope(), clock = FakeClock())
        client.setState(ProgressorConnectionState.Connected)
        val activity = ActivityRecorder()
        val service = ServiceRecorder()
        val session = RunnerSession(
            plan = SessionPlan(sets = listOf(SetPlan(repsPerSide = 2)), handMode = HandMode.bothHands,
                holdSeconds = 1, restSeconds = 20, leadInSeconds = 0, waitForReleaseBeforeRest = true),
            routineName = "Activity lifecycle", device = device, scope = inertScope(), clock = FakeClock(),
            activity = activity, service = service,
        )
        session.begin()
        for (index in 1..120) session.send(RunnerEvent.Sample(ForceSample(10.0, (index * 12_500).toUInt())))
        assertTrue(session.snapshot.phase is RunnerPhase.Releasing)
        assertEquals(SessionActivityPhase.releasing, activity.states.last().phase)
        assertNull(activity.states.last().endsAtEpochMillis)
        session.send(RunnerEvent.Pause)
        assertEquals(SessionActivityPhase.paused, activity.states.last().phase)
        assertNull(activity.states.last().endsAtEpochMillis)
        session.send(RunnerEvent.Resume)
        for (index in 121..160) session.send(RunnerEvent.Sample(ForceSample(0.0, (index * 12_500).toUInt())))
        assertEquals(SessionActivityPhase.resting, activity.states.last().phase)
        assertTrue(activity.states.last().endsAtEpochMillis!! - System.currentTimeMillis() > 18_000)
        val published = activity.states.size
        session.send(RunnerEvent.Abort)
        assertEquals(1, activity.ends, "End on the summary, before save/discard")
        assertEquals(1, service.ends)
        assertEquals(published, activity.states.size, "Finish must not publish a phantom rest")

        val commands = client.commands.size
        session.startIfReady(StreamStartCause.foreground)
        session.connectionChanged(true)
        session.wakeStream()
        session.tickNow()
        assertEquals(commands, client.commands.size)
        assertEquals(1, activity.starts)
        assertEquals(1, service.starts)
        assertEquals(published, activity.states.size)
        session.end()
        assertEquals(1, activity.ends)
        assertEquals(1, service.ends)
    }
}

File: Tests/RunnerActivityLifecycleTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

@MainActor
final class RunnerActivityLifecycleTests: XCTestCase {
    private final class Recorder: RunnerActivityPublishing {
        var isRunning = false
        var starts = 0
        var ends = 0
        var states: [SessionActivity.ContentState] = []
        func start(routineName: String, plannedReps: Int, setCount: Int,
                   state: SessionActivity.ContentState) {
            starts += 1
            isRunning = true
            states.append(state)
        }
        func update(_ state: SessionActivity.ContentState) async {
            if isRunning { states.append(state) }
        }
        func end() async {
            guard isRunning else { return }
            isRunning = false
            ends += 1
        }
    }

    func testReleaseRestPauseAndFinishPublishTheirActualLifecycle() async throws {
        var draft = RoutineDraft.blank(named: "Activity lifecycle")
        draft.plan.sets = [SetPlan(repsPerSide: 2)]
        draft.plan.handMode = .bothHands
        draft.plan.leadInSeconds = 0
        draft.plan.holdSeconds = 1
        draft.plan.restSeconds = 20
        draft.plan.waitForReleaseBeforeRest = true
        let client = RecordingProgressorClient()
        let device = DeviceStore(client: client)
        client.setState(.connected)
        let activity = Recorder()
        let session = RunnerSession(template: SessionTemplate(draft: draft, sortIndex: 0),
                                    device: device, liveActivity: activity)
        session.begin()
        defer { session.end() }
        for index in 1...120 {
            session.send(.sample(ForceSample(kg: 10, deviceMicros: UInt32(index * 12_500))))
        }
        guard case .releasing = session.snapshot.phase else { return XCTFail("Expected release wait") }
        await Task.yield()
        XCTAssertEqual(activity.states.last?.phase, .releasing)
        XCTAssertNil(activity.states.last?.endsAt, "LET GO waits for force, without a rest deadline")
        session.send(.pause)
        await Task.yield()
        XCTAssertEqual(activity.states.last?.phase, .paused)
        XCTAssertNil(activity.states.last?.endsAt)
        session.send(.resume)
        // The real release debounce is sample-driven too.
        for index in 121...160 {
            session.send(.sample(ForceSample(kg: 0, deviceMicros: UInt32(index * 12_500))))
        }
        await Task.yield()
        XCTAssertEqual(activity.states.last?.phase, .resting)
        let restDeadline = try XCTUnwrap(activity.states.last?.endsAt)
        XCTAssertLessThanOrEqual(restDeadline.timeIntervalSinceNow, 20.01,
                                 "The Live Activity must not inherit a rounded extra second")
        XCTAssertGreaterThan(restDeadline.timeIntervalSinceNow, 18,
                             "The full rest clock is anchored when the load releases")
        let published = activity.states.count
        session.send(.abort)
        await Task.yield()
        XCTAssertEqual(activity.ends, 1, "End on the summary, before save/discard")
        XCTAssertEqual(activity.states.count, published, "Finish must not publish a phantom rest")

        let commands = client.commands.count
        session.startIfReady(cause: .foreground)
        session.connectionChanged(isConnected: true)
        session.wakeStream()
        session.send(.tick)
        await Task.yield()
        XCTAssertEqual(client.commands.count, commands, "A summary never restarts the gauge")
        XCTAssertEqual(activity.starts, 1)
        XCTAssertEqual(activity.states.count, published)
    }
    func testActivityUsesSelectedWeightUnitWithoutChangingTargets() async throws {
        var draft = RoutineDraft.blank(named: "Pound targets")
        draft.plan.sets = [SetPlan(repsPerSide: 2, targetLoKg: 10, targetHiKg: 15)]
        draft.plan.handMode = .bothHands
        draft.plan.leadInSeconds = 0
        let client = RecordingProgressorClient()
        let device = DeviceStore(client: client)
        client.setState(.connected)
        let activity = Recorder()
        let session = RunnerSession(template: SessionTemplate(draft: draft, sortIndex: 0),
                                    device: device, liveActivity: activity)
        session.weightUnit = .lb
        session.begin()
        defer { session.end() }
        XCTAssertEqual(activity.states.last?.displayWeightUnit, .lb)
        XCTAssertEqual(activity.states.last?.targetLoKg, 10)
        XCTAssertEqual(activity.states.last?.targetHiKg, 15)
        let published = activity.states.count
        session.weightUnit = .kg
        await Task.yield()
        XCTAssertEqual(activity.states.count, published + 1)
        XCTAssertEqual(activity.states.last?.displayWeightUnit, .kg)
        XCTAssertEqual(session.snapshot.targetBand, 10...15)
    }

}
